src/topology/connected.lean

```diff
@@ -353,6 +353,79 @@ begin
   exact hc.image _ (continuous_apply _).continuous_on
 end
 
+lemma sigma.is_connected_iff [Π i, topological_space (π i)] {s : set (Σ i, π i)} :
+  is_connected s ↔ ∃ i t, is_connected t ∧ s = sigma.mk i '' t :=
+begin
+  refine ⟨λ hs, _, _⟩,
+  { obtain ⟨⟨i, x⟩, hx⟩ := hs.nonempty,
+    have : s ⊆ set.range (sigma.mk i),
+    { have h : set.range (sigma.mk i) = sigma.fst ⁻¹' {i}, by { ext, simp },
+      rw h,
+      exact is_preconnected.subset_left_of_subset_union
+        (is_open_sigma_fst_preimage _) (is_open_sigma_fst_preimage {x | x ≠ i})
+        (set.disjoint_iff.2 $ λ x hx, hx.2 hx.1)
+        (λ y hy, by simp [classical.em]) ⟨⟨i, x⟩, hx, rfl⟩ hs.2 },
+    exact ⟨i, sigma.mk i ⁻¹' s,
+      hs.preimage_of_open_map sigma_mk_injective is_open_map_sigma_mk this,
+      (set.image_preimage_eq_of_subset this).symm⟩ },
+  { rintro ⟨i, t, ht, rfl⟩,
+    exact ht.image _ continuous_sigma_mk.continuous_on }
+end
+
+lemma sigma.is_preconnected_iff [hι : nonempty ι] [Π i, topological_space (π i)]
```

Comment:
Funny. I would've expected that you prove this one first, and use it to prove the lemma above about connectedness. Wouldn't that lead to a shorter total proof?

Reply:
Not really no, because `sigma.is_connected_iff` crucially uses that `s` is nonempty to decide which side it comes from. So I could case on `s` being empty or nonempty but at that point I already know that the set it's coming is connected and not merely preconnected. Also note the `[hι : nonempty ι]`.